# PC1 - Grupo 2 (Julia)

#### Integrantes
- GARCIA RODRIGUEZ, EMILIO ALONSO
- PADILLA AQUISE, ALESSANDRO PIERO
- RIEGA NUÑEZ, GABRIEL ANTONIO FERMIN
- SALAMANCA FERNANDEZ, LUCAS PABLO
- SILVA ANDUJAR, NICOLAS

## 1. Data analysis

In [16]:
using Pkg
#Pkg.add("Plots")
#Pkg.add("Random")
#Pkg.add("GLM")
#Pkg.add("DataFrames")
#Pkg.add("LaTeXStrings")
#Pkg.add("Statistics") 
#Pkg.add("StatsBase")
#Pkg.add("CSV")
#Pkg.add("CategoricalArrays")

   Resolving package versions...
    Updating `~/.julia/environments/v1.10/Project.toml`
  [324d7699] + CategoricalArrays v0.10.8
  No Changes to `~/.julia/environments/v1.10/Manifest.toml`


In [1]:
using Plots
using Random
using GLM
using DataFrames
using LaTeXStrings
using Statistics
using StatsBase

1.Import the data set. Make sure the column names are imported as intended variable

In [2]:
using CSV

data = CSV.read("wage2015_subsample_inference.csv", DataFrame)

first(data,5)

5×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg    ⋯
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float6 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │       10   9.61538  2.26336      1.0      0.0      0.0      0.0      1. ⋯
   2 │       12  48.0769   3.8728       0.0      0.0      0.0      0.0      1.
   3 │       15  11.0577   2.40313      0.0      0.0      1.0      0.0      0.
   4 │       18  13.9423   2.63493      1.0      0.0      0.0      0.0      0.
   5 │       19  28.8462   3.36198      1.0      0.0      0.0      0.0      1. ⋯
                                                              14 columns omitted

2.Are there missing values? Display the number of missings (if any) of each variable in the data set.

In [3]:
# Find number of missing values for each column 
missing_values = map(col -> sum(ismissing, col), eachcol(data))

# Handle missing values a dataframe
missing_df = DataFrame(Column = names(data), MissingCount = missing_values)

println(missing_df)

# Check for missing values 
if all(count -> count == 0, missing_values)
    println("There are no missing values")
else
    println("There are missing values")
end


21×2 DataFrame
 Row │ Column    MissingCount 
     │ String    Int64        
─────┼────────────────────────
   1 │ rownames             0
   2 │ wage                 0
   3 │ lwage                0
   4 │ sex                  0
   5 │ shs                  0
   6 │ hsg                  0
   7 │ scl                  0
   8 │ clg                  0
   9 │ ad                   0
  10 │ mw                   0
  11 │ so                   0
  12 │ we                   0
  13 │ ne                   0
  14 │ exp1                 0
  15 │ exp2                 0
  16 │ exp3                 0
  17 │ exp4                 0
  18 │ occ                  0
  19 │ occ2                 0
  20 │ ind                  0
  21 │ ind2                 0
There are no missing values


3.Report descriptive statistics of the variables (mean, standard deviation, percentiles, etc.). Interpret your results.

In [4]:
describe(data)


21×7 DataFrame
 Row │ variable  mean          min        median       max          nmissing   ⋯
     │ Symbol    Float64       Real       Float64      Real         Int64      ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ rownames  15636.3        10        15260.0       32643              0   ⋯
   2 │ wage         23.4104      3.02198     19.2308      528.846          0
   3 │ lwage         2.97079     1.10591      2.95651       6.2707         0
   4 │ sex           0.444466    0.0          0.0           1.0            0
   5 │ shs           0.023301    0.0          0.0           1.0            0   ⋯
   6 │ hsg           0.243883    0.0          0.0           1.0            0
   7 │ scl           0.278058    0.0          0.0           1.0            0
   8 │ clg           0.31767     0.0          0.0           1.0            0
  ⋮  │    ⋮           ⋮            ⋮           ⋮            ⋮          ⋮       ⋱
  15 │ exp2          3.01893     0.0          1.0          22.09           0   ⋯
  16 │ exp3          8.23587     0.0          1.0         103.823          0
  17 │ exp4         25.118       0.0          1.0         487.968          0
  18 │ occ        5310.74       10.0       4040.0      100000.0            0
  19 │ occ2         11.6709      1           13.0          22              0   ⋯
  20 │ ind        6629.15      370.0       7370.0      100000.0            0
  21 │ ind2         13.3169      2           14.0          22              0
                                                     1 column and 6 rows omitted

In [7]:
function calculate_quartiles(df)
    quartiles = DataFrame(
        Variable = names(df),
        Q1 = quantile.(eachcol(df), Ref(0.25)),
        Q2 = quantile.(eachcol(df), Ref(0.50)),  # Median
        Q3 = quantile.(eachcol(df), Ref(0.75))
    )
    return quartiles
end

quartiles_summary = calculate_quartiles(data)

println(quartiles_summary)

21×4 DataFrame
 Row │ Variable  Q1          Q2           Q3          
     │ String    Float64     Float64      Float64     
─────┼────────────────────────────────────────────────
   1 │ rownames  7132.5      15260.0      24337.8
   2 │ wage        13.4615      19.2308      27.7778
   3 │ lwage        2.59984      2.95651      3.32424
   4 │ sex          0.0          0.0          1.0
   5 │ shs          0.0          0.0          0.0
   6 │ hsg          0.0          0.0          0.0
   7 │ scl          0.0          0.0          1.0
   8 │ clg          0.0          0.0          1.0
   9 │ ad           0.0          0.0          0.0
  10 │ mw           0.0          0.0          1.0
  11 │ so           0.0          0.0          1.0
  12 │ we           0.0          0.0          0.0
  13 │ ne           0.0          0.0          0.0
  14 │ exp1         5.0         10.0         21.0
  15 │ exp2         0.25         1.0          4.41
  16 │ exp3         0.125        1.0          9.261
  17 │ exp

In [9]:
#Std deviation

# Function that calculates the std deviation for each column
function calculate_standard_deviation(df::DataFrame)
    std_values = Dict{Symbol, Any}()
    
    for col in names(df)
        # Change column name to symbol
        col_sym = Symbol(col)
            
        # Calculate std deviation
        std_val = std(df[!, col_sym])
        std_values[col_sym] = std_val
        
    end
    return std_values
end

std_dict = calculate_standard_deviation(data)
println("Standard deviation for each variable:")
println(std_dict)

Standard deviation for each variable:
Dict{Symbol, Any}(:so => 0.45676058983904344, :hsg => 0.4294649573160264, :mw => 0.438464105331998, :wage => 21.003016115654056, :exp1 => 10.609465081743176, :occ => 11874.35608007378, :sex => 0.4969546570483574, :ind2 => 5.701019018248305, :ne => 0.419432173656693, :rownames => 9700.744394538977, :occ2 => 6.966684461108375, :lwage => 0.5703847829926473, :we => 0.4116346228750478, :exp2 => 4.000904430517199, :exp3 => 14.488961873778786, :exp4 => 53.53022497424143, :shs => 0.15087231530138967, :scl => 0.4480857588192326, :ad => 0.3439729691578773, :ind => 5333.443992032589, :clg => 0.4656155413671081)


#### wage:
The average wage in the sample is 23.41, with a standard deviation of 21.03, reflecting significant variability in earnings. Wages range from a minimum of 3.02 to a maximum of 528.85, indicating a broad spectrum of income levels. The bottom 25th percentile earn less than 13.46, while the median wage is 19.23, and the top 25% earn more than 27.78. The fact that the median is 19.23 suggests the presence of a number of extremely high wage values, which likely skews the distribution.

#### lwage:
The logarithm of wage has a mean of 2.97 and a standard deviation of 0.57, suggesting that the distribution of wages in its logarithmic scale is more symmetric than in its original form. The logarithmic wage values range from a minimum of 1.11 to a maximum of 6.27, with the top 25% of the sample having a log wage greater than 3.32. Also, it has a median of 2.96, that suggest a lower presence of extremely high or low wages in the sample. 

#### sex:
Although sex is a dummy variable, we can see that 44% of the sample are females, while the mode is "male." The standard deviation of 0.5 suggests an almost balanced distribution between males and females.

#### shs:
The "some high school" variable shows that only 2.3% of the sample have some incomplete high school education, which represents a small fraction of the total population. 

#### hsg:
About 24% of the sample has graduated from high school, according to the mean of this variable. The standard deviation of 0.43 reflects moderate diversity in the education levels within the sample, with a significant portion not having attained a college degree. 

#### scl:
Approximately 28% of the respondents have attended some college without necessarily graduating, as indicated by the mean. This suggests that a considerable segment of the population has received some tertiary education, although they may not have completed their studies.

#### clg:
The data shows that 32% of the sample holds a college degree, as reflected in the mean of this variable. This implies that nearly one-third of the sample has completed a college education, which is significant for evaluating human capital.

#### ad:
About 14% of the individuals in the sample have an advanced degree (master's or doctorate), according to the mean of this variable. This relatively low percentage highlights that advanced education is less common compared to basic college education.

#### mw:
The Midwest region represents 26% of the sample, as indicated by the mean of this variable. This suggests that this region has significant representation within the dataset.

#### so:
The South is the most represented region, with 30% of the sample residing in this area, as shown by the mean of this variable. The higher presence of the South may reflect certain geographical trends in the sample.

#### we:
The West region accounts for 22% of the sample, according to the mean. This proportion suggests a relatively equitable distribution among the country's regions, although with less representation compared to the South.

#### ne:
The Northeast region represents 23% of the sample, as indicated by the mean of this variable. This proportion, similar to the other regions, shows a relatively balanced geographical distribution in the dataset.

#### exp1:
The average work experience is 13.76 years, with a standard deviation of 10.61, indicating high variability in the number of years worked among the individuals in the sample. The median is 10 years, meaning that half of the sample has 10 years or less of work experience, while the maximum is 47 years. Also, the mode of work experience is 5.

#### exp2:
The first polynomial of work experience has a mean of 3.02 years, with a standard deviation of 4.00. This suggests that, for certain specific aspects of work experience, most of the sample has relatively low experience.

#### exp3:
The second polynomial of work experience, exp3, has a mean of 8.24 years and a standard deviation of 14.49, indicating a wide variability.

#### exp4:
The third polynomial of work experience, exp4, shows a significantly higher mean of 25.12 years, with a standard deviation of 53.53, again reflecting a high variability.

4.How many women with a college graduate degree (clg) or above have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [10]:
# Calculate 75th percentile based on wage
percentile_75_wage = quantile(skipmissing(data.wage), 0.75)

# Filter educated women with a salary belonging to the top 25% 
rich_ed_women = filter(row -> (row.sex == 1) && 
                             ((row.clg == 1) || (row.ad == 1)) && 
                             (row.wage >= percentile_75_wage), data)

rich_ed_women

419×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg    ⋯
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float6 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │       19   28.8462  3.36198      1.0      0.0      0.0      0.0      1. ⋯
   2 │      191   42.3077  3.74497      1.0      0.0      0.0      0.0      0.
   3 │      232   41.2088  3.71865      1.0      0.0      0.0      0.0      1.
   4 │      319  100.0     4.60517      1.0      0.0      0.0      0.0      0.
   5 │      563   33.6538  3.51613      1.0      0.0      0.0      0.0      1. ⋯
   6 │      842   36.0577  3.58512      1.0      0.0      0.0      0.0      1.
   7 │      865   31.25    3.44202      1.0      0.0      0.0      0.0      1.
   8 │      892   28.8462  3.36198      1.0      0.0      0.0      0.0      1.
  ⋮  │    ⋮         ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮   ⋱
 413 │    31956   28.8462  3.36198      1.0      0.0      0.0      0.0      0. ⋯
 414 │    32160   50.3635  3.91927      1.0      0.0      0.0      0.0      0.
 415 │    32251   48.0769  3.8728       1.0      0.0      0.0      0.0      0.
 416 │    32314   30.1455  3.40604      1.0      0.0      0.0      0.0      1.
 417 │    32419   29.9145  3.39834      1.0      0.0      0.0      0.0      1. ⋯
 418 │    32567   48.0769  3.8728       1.0      0.0      0.0      0.0      1.
 419 │    32596   45.5466  3.81874      1.0      0.0      0.0      0.0      1.
                                                 14 columns and 404 rows omitted

In [11]:
println("Number of women with a college graduate degree or above in the top 25% richest:",size(rich_ed_women,1))

Number of women with a college graduate degree or above in the top 25% richest:419


5.How many men with a high school graduate degree (hsg) or below have a wage corresponding to the 25% richest of the sample? Report the dataframe corresponding to this conditions and the result.

In [12]:
# Filter men with high school or below
rich_uned_men = filter(row-> (row.sex == 0) &&
                            ((row.hsg == 1) || (row.shs == 1)) &&
                            (row.wage >= percentile_75_wage), data)

rich_uned_men

118×21 DataFrame
 Row │ rownames  wage      lwage    sex      shs      hsg      scl      clg    ⋯
     │ Int64     Float64   Float64  Float64  Float64  Float64  Float64  Float6 ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │      113   27.8846  3.32808      0.0      0.0      1.0      0.0      0. ⋯
   2 │      276   28.8462  3.36198      0.0      0.0      1.0      0.0      0.
   3 │      467   28.8462  3.36198      0.0      0.0      1.0      0.0      0.
   4 │      858   28.8462  3.36198      0.0      0.0      1.0      0.0      0.
   5 │      876   29.7143  3.39163      0.0      0.0      1.0      0.0      0. ⋯
   6 │     1065   27.7778  3.32424      0.0      0.0      1.0      0.0      0.
   7 │     1359   35.5769  3.5717       0.0      0.0      1.0      0.0      0.
   8 │     1870   36.3248  3.5925       0.0      0.0      1.0      0.0      0.
  ⋮  │    ⋮         ⋮         ⋮        ⋮        ⋮        ⋮        ⋮        ⋮   ⋱
 112 │    31922   33.6538  3.51613      0.0      0.0      1.0      0.0      0. ⋯
 113 │    32075   43.2692  3.76744      0.0      1.0      0.0      0.0      0.
 114 │    32191   33.6538  3.51613      0.0      0.0      1.0      0.0      0.
 115 │    32254   40.8654  3.71028      0.0      0.0      1.0      0.0      0.
 116 │    32281   31.25    3.44202      0.0      0.0      1.0      0.0      0. ⋯
 117 │    32321   33.6538  3.51613      0.0      0.0      1.0      0.0      0.
 118 │    32631   32.967   3.49551      0.0      0.0      1.0      0.0      0.
                                                 14 columns and 103 rows omitted

In [13]:
println("Number of men with a high school degree or below in the top 25% richest:", size(rich_uned_men,1))

Number of men with a high school degree or below in the top 25% richest:118


6.Create two dataframes. One containing only the log(wage) and the other containig every variable of the data set but the wage related variable. 


In [14]:
# DataFrame for'lwage'
log_w = data[:, [:lwage]]

# DataFrame for the rest of the columns except 'rownames' 'wage' y 'lwage'
regs = select(data, Not([:rownames, :wage, :lwage]))

display(log_w)
display(regs)


5150×1 DataFrame
  Row │ lwage   
      │ Float64 
──────┼─────────
    1 │ 2.26336
    2 │ 3.8728
    3 │ 2.40313
    4 │ 2.63493
    5 │ 3.36198
    6 │ 2.46222
    7 │ 2.95651
    8 │ 2.95651
  ⋮   │    ⋮
 5144 │ 3.17966
 5145 │ 2.62801
 5146 │ 2.69255
 5147 │ 3.13883
 5148 │ 3.64966
 5149 │ 3.49551
 5150 │ 2.85115
5135 rows omitted

5150×18 DataFrame
  Row │ sex      shs      hsg      scl      clg      ad       mw       so      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    2 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    3 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    4 │     1.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
    5 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
    6 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
    7 │     1.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
    8 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5144 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0 ⋯
 5145 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5146 │     0.0      0.0      0.0      0.0      1.0      0.0      0.0      0.0
 5147 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      0.0
 5148 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0 ⋯
 5149 │     0.0      0.0      1.0      0.0      0.0      0.0      0.0      0.0
 5150 │     0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0
                                                10 columns and 5135 rows omitted

## 2. Data Wrangling

7.Make an array for our y variable, which will be the logarithm of wage (lwage column). 

In [16]:
#This has aleady been solved in question 6, as there is no need to create a specific array in Julia

size(log_w)

(5150, 1)

8.Make three new arrays for our predictors:

8.1. The basic model will have the columns sex hsg scl clg ad so we ne exp1 occ2 ind2. Make sure to convert occ2 and ind2 to dummies and to drop the first dummy value to prevent multicolinearity. This means:

In [17]:
# Categories for occ2
sort(unique(data.occ2))


22-element Vector{Int64}:
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
  ⋮
 14
 15
 16
 17
 18
 19
 20
 21
 22

In [27]:
# Make occ2 a dummy:
basic_model = regs
for i in 2:22
    col_name = Symbol("occ2_$i")
    basic_model = transform(basic_model, :occ2 => ByRow(x -> x == i ? 1 : 0) => col_name)
end

println(basic_model)


In [28]:
# Categories for ind2
sort(unique(data.ind2))

21-element Vector{Int64}:
  2
  3
  4
  5
  6
  7
  8
  9
 10
 11
  ⋮
 14
 15
 16
 17
 18
 19
 20
 21
 22

In [29]:
# Make occ2 a dummy:
for i in 2:21
    col_name = Symbol("ind2_$i")
    basic_model = transform(basic_model, :ind2 => ByRow(x -> x == i ? 1 : 0) => col_name)
end

basic_model
ncol(basic_model)

59

In [30]:
#Basic Model

select!(basic_model, Not(["shs", "mw", "occ", "ind", "exp2", "exp3", "exp4"]))
basic_model

5150×52 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
                                                44 columns and 5135 rows omitted

8.2. The flexible model will have the same columns, and will also include polynomials for experience (exp2 exp3 exp4), as well as the interactions between all experience variables and other variables except for sex. This means:

In [32]:
#Flexible Model 

flex_model = basic_model

insertcols!(flex_model, :exp2 => regs.exp2, :exp3 => regs.exp3, :exp4 => regs.exp4)

flex_model



5150×55 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
                                                47 columns and 5135 rows omitted

In [38]:
# Flexible model - interactions

exp_vars = [:exp1, :exp2, :exp3, :exp4]

other_vars = [
    :hsg, :scl, :clg, :ad, :so, :we, :ne,  
    :occ2, :ind2,  
    :occ2_2, :occ2_3, :occ2_4, :occ2_5, :occ2_6, :occ2_7, :occ2_8, :occ2_9, 
    :occ2_10, :occ2_11, :occ2_12, :occ2_13, :occ2_14, :occ2_15, :occ2_16, 
    :occ2_17, :occ2_18, :occ2_19, :occ2_20, :occ2_21, :occ2_22,
    :ind2_2, :ind2_3, :ind2_4, :ind2_5, :ind2_6, :ind2_7, :ind2_8, :ind2_9,
    :ind2_10, :ind2_11, :ind2_12, :ind2_13, :ind2_14, :ind2_15, :ind2_16,
    :ind2_17, :ind2_18, :ind2_19, :ind2_20, :ind2_21
]

# Create interactions
for exp in exp_vars
    for var in other_vars
        # New column name
        interaction_name = Symbol("$(exp)_x_$(var)")
        
        flex_model[!, interaction_name] = flex_model[!, exp] .* flex_model[!, var]
    end
end

flex_model

5150×255 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
                                               247 columns and 5135 rows omitted

8.3. The extra-flexible model will include all two-way interactions between variables, except for sex. 

In [42]:
# Extra flexible model

extra_flex_model = flex_model


# Variables without "sex" 
vars_to_interact = [
    :exp1, :exp2, :exp3, :exp4, :hsg, :scl, :clg, :ad, :so, :we, :ne,
    :occ2, :ind2, :occ2_2, :occ2_3, :occ2_4, :occ2_5, :occ2_6, :occ2_7, :occ2_8, 
    :occ2_9, :occ2_10, :occ2_11, :occ2_12, :occ2_13, :occ2_14, :occ2_15, :occ2_16, 
    :occ2_17, :occ2_18, :occ2_19, :occ2_20, :occ2_21, :occ2_22, 
    :ind2_2, :ind2_3, :ind2_4, :ind2_5, :ind2_6, :ind2_7, :ind2_8, :ind2_9, 
    :ind2_10, :ind2_11, :ind2_12, :ind2_13, :ind2_14, :ind2_15, :ind2_16, 
    :ind2_17, :ind2_18, :ind2_19, :ind2_20, :ind2_21
]

# Create two way interactions
for i in 1:length(vars_to_interact)
    for j in (i+1):length(vars_to_interact)  # begin from `i+1` to avoid duplicates
        var1 = vars_to_interact[i]
        var2 = vars_to_interact[j]

        # Name of new interactions
        interaction_name = Symbol("$(var1)_x_$(var2)")

        # Create interactions
        extra_flex_model[!, interaction_name] = extra_flex_model[!, var1] .* extra_flex_model[!, var2]
    end
end

extra_flex_model


5150×1486 DataFrame
  Row │ sex      hsg      scl      clg      ad       so       we       ne      ⋯
      │ Float64  Float64  Float64  Float64  Float64  Float64  Float64  Float64 ⋯
──────┼─────────────────────────────────────────────────────────────────────────
    1 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    2 │     0.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    3 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    4 │     1.0      0.0      0.0      0.0      1.0      0.0      0.0      1.0
    5 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0 ⋯
    6 │     1.0      0.0      0.0      1.0      0.0      0.0      0.0      1.0
    7 │     1.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
    8 │     0.0      1.0      0.0      0.0      0.0      0.0      0.0      1.0
  ⋮   │    ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮        ⋮    ⋱
 5144 │     1.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0 ⋯
 5145 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5146 │     0.0      0.0      0.0      1.0      0.0      0.0      1.0      0.0
 5147 │     1.0      0.0      1.0      0.0      0.0      0.0      1.0      0.0
 5148 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0 ⋯
 5149 │     0.0      1.0      0.0      0.0      0.0      0.0      1.0      0.0
 5150 │     0.0      0.0      0.0      0.0      1.0      0.0      1.0      0.0
                                              1478 columns and 5135 rows omitted